lib/AsyncGenerator.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace Amp;
+
+/**
+ * @template TValue
+ * @template TKey
+ */
+final class AsyncGenerator implements Stream
```

Comment:
I really dislike the name of this class. From this name I expected something like normal PHP's Generator except Asynchronous. Meaning it should be able to both emit values AND receive them using something similar to `Generator::send();`:

```
$recievedValue = yield $yield($emittedValue);
```

Since this isn't the case I highly recommend renaming this class to something more appropriate.

Of course this kind of stream would have no backpressure since it needs to receive a value before it can continue execution. I'm not necessarily suggesting for you to add it since the use case for it is rare, but the name here is very confusing for me.

Reply:
Very much open to name suggestions. I agree that we should keep `AsyncGenerator` as the name of a class that does what you describe, in case a use does arise.

Reply:
I admit that naming this class is a bit tough. I do like the name Producer which is currently used for async iterators. So possibly `StreamProducer` to make it clear it's related to the new streams feature?

Reply:
Naming is hard and I wanted to try doing it, so I actually implemented it as an async generator. 🤓

Reply:
@trowski 🥇

Reply:
I seem to have an ordering issue to work out… guess it wasn't as simple as I hoped.

Reply:
Now it should work as expected! 😁

Reply:
I don't think we need `AsyncGenerator`, I'm sure we can find a better name for it.

Reply:
Why not? Maybe there's something interesting that could be implemented with it. And it still works the same as `Producer` if you're not interested in send/throw.